docs/manual/src/internals/rendering_foreign_bindings.md

```diff
@@ -0,0 +1,82 @@
+# Rendering Foreign Bindings
+
+This document details the general system that UniFFI uses to render the foreign bindings code.
+
+## The Askama template engine
+
+Our foreign bindings generation is based off the [Askama](https://djc.github.io/askama/) template rendering engine.
+Askama uses a compile-time macro system that allows the template code to use Rust types directly, calling their methods
+passing them to normal Rust functions.
+
+## Type matching
+
+One of the main sources of complexity when generating the bindings is handling types.  UniFFI supports a large number of
+types, each of which corresponds to a variant of the [`Type enum`](./api/uniffi_bindgen/interface/types/enum.Type.html).
+At one point there was a fairly large number of "mega-match" functions, each one matching against all `Type` variants.
```

Comment:
May be worth mentioning the [State pattern][1] here.

[1]: https://www.gofpatterns.com/behavioral/patterns/state-pattern.php